### PR TITLE
feat: Add `ValidatingValinorParamConverter`

### PR DESCRIPTION
**Pull request by gertjuhh**

### CODE REVIEW

File: src/Symfony/Request/ParamConverter/ValinorParamConverter.php [hunk 1]
```diff
@@ -46,4 +43,32 @@ protected function getMapperBuilder(): MapperBuilder
...
+            foreach ($flattenedMessages as $message) {
+                $errors[\str_replace(\sprintf('.%s', $message->name()), '', $message->path())] = $message->__toString();
+            }
+
+            throw new JsonApiProblem(
+                'Invalid Request',
+                'Invalid data provided.',
+                Response::HTTP_UNPROCESSABLE_ENTITY,
+                ['errors' => $errors]
+            );
```
Comment (frankverhoeven) [inline src/Symfony/Request/ParamConverter/ValinorParamConverter.php:81]: Why did you move this logic out of `JsonApiProblem::fromValinorMappingError()`?

**Review by frankverhoeven — COMMENTED** [review]
Note that advanced validation that can not be accomplished with a simple value object and/or collection, should _probably_ not be part of the request mapping. It should be part of your domain instead.
The issue that may arise here is that you're passing data to your domain of which you assume that it is valid. Assume, because it is not validated by the domain itself, but by the request mapping. If you ever introduce another path to provide this data to the domain, eg a new API endpoint or a Kafka consumer, it may not contain the same mapping and thus miss this validation. So simply put; because your domain doesn't perform the validation, it is possible to provide your domain with invalid data.
Taking that into consideration, I'm not entirely sure if we should add this change to the package.

**Comment by gertjuhh** [comment]
> Note that advanced validation that can not be accomplished with a simple value object and/or collection, should _probably_ not be part of the request mapping. It should be part of your domain instead. The issue that may arise here is that you're passing data to your domain of which you assume that it is valid. Assume, because it is not validated by the domain itself, but by the request mapping. If you ever introduce another path to provide this data to the domain, eg a new API endpoint or a Kafka consumer, it may not contain the same mapping and thus miss this validation. So simply put; because your domain doesn't perform the validation, it is possible to provide your domain with invalid data. Taking that into consideration, I'm not entirely sure if we should add this change to the package.

What I did in the store-branding-service was add DTOs which were used to map the requests. These DTOs should not be used for anything other then wrapping input and should therefor be safe to be validated by the request mapper.

**Comment by frankverhoeven** [comment]
From what I can see, it is exactly what my comment is about. Constraints like these https://github.com/MyOnlineStore/store-branding-service/blob/51ea177ae8047fddffcc0815e0d074e5fd90eccb/src/Dto/Colors.php#L15-L22 are really part of your domain logic, not of the request mapping. The `Colors` DTO can also be instantiated without triggering the validation, at which point it may contain invalid data. Thus strictly speaking it can't be trusted and must be re-validated by your domain.

To illustrate the reasoning:

image

**Comment by gertjuhh** [comment]
I think you're illustration somewhat over simplifies the situation.
The DTO (and it's valiation rules) are part of the domain.

> The Colors DTO can also be instantiated without triggering the validation

In theory, yes. But it shouldn't, and that's what reviews are for.
Although I get the reasoning behind your feedback, I think it's a hypothetical problem rather then a real world issue.

**Comment by frankverhoeven** [comment]
> I think you're illustration somewhat over simplifies the situation.

It is indeed a simplification to explain the theory.

> Although I get the reasoning behind your feedback, I think it's a hypothetical problem rather then a real world issue.

It is a very realistic and practical issue that we're dealing with every day. The problem starts with the fact that you have an [anemic domain model](https://martinfowler.com/bliki/AnemicDomainModel.html) (entity without domain logic). Because the logic must live somewhere, it is moved to a thick service layer, glued together with fat controllers. Sounds exactly like the current state of our monolith to me.

In store branding, exactly the same is happening, but on a smaller scale. An anemic entity without logic is used. It is validated in the request mapping. And a controller ties it together. But because your entity does not contain any logic, it can be used and abused in any way possible. The actual domain logic doesn't reside in the domain layer. Expecting this to be caught by a review is quite optimistic, and working on the monolith has proven that that does not work.

**Comment by gertjuhh** [comment]
> > Although I get the reasoning behind your feedback, I think it's a hypothetical problem rather then a real world issue.
> 
> It is a very realistic and practical issue that we're dealing with every day. 

You're completely taking my reply out of context, I was merely stating that misuse of a DTO should be captured during the review process.

> The problem starts with the fact that you have an [anemic domain model](https://martinfowler.com/bliki/AnemicDomainModel.html) (entity without domain logic). Because the logic must live somewhere, it is moved to a thick service layer, glued together with fat controllers. Sounds exactly like the current state of our monolith to me.
> 
> In store branding, exactly the same is happening, but on a smaller scale. An anemic entity without logic is used. It is validated in the request mapping. And a controller ties it together. But because your entity does not contain any logic, it can be used and abused in any way possible. The actual domain logic doesn't reside in the domain layer. Expecting this to be caught by a review is quite optimistic, and working on the monolith has proven that that does not work.

Thanks for the feedback, I understand you see room for improvement in the branding service.
I'd be more then happy to get together and hear you feedback, but I don't think a PR review is the place to discuss this.

**Comment by frankverhoeven** [comment]
> You're completely taking my reply out of context, I was merely stating that misuse of a DTO should be captured during the review process.

Like I said, history has shown us that that won't work. When the original authors leave and others without the same domain knowledge have to take over, things will go downhill. I don't think that that is taking anything out of context, it is a mere reaction to your statement that things are hypothetical and theoretical, while we are dealing with said issues everyday.

> but I don't think a PR review is the place to discuss this.

If we go back to the start of the discussion, I told you I'm not sure whether we should embrace this change because of the potential misuse that it allows for. Looking at the use-case that it was introduced for, I've found exactly what I was worried about. So yes, I do think this discussion is relevant, because it determines the outcome of this PR. Note that the discussion is not about potential changes for store branding to me, but store branding is used as example to illustratie the pitfalls of the change.
